src/libraries/System.Private.Uri/src/System/PercentEncodingHelper.cs

```diff
@@ -0,0 +1,157 @@
+            Debug.Assert(length >= 3);
+            Debug.Assert(input[0] == '%');
+            Debug.Assert(UriHelper.EscapedAscii(input[1], input[2]) != Uri.c_DummyChar);
+            Debug.Assert(UriHelper.EscapedAscii(input[1], input[2]) >= 128);
+
+            uint fourByteBuffer = 0;
+            int bytesLeftInBuffer = 0;
+            var fourByteSpan = new ReadOnlySpan<byte>(&fourByteBuffer, sizeof(uint));
+
+            int totalCharsConsumed = 0;
+            int charsToCopy = 0;
+            int bytesConsumed = 0;
+
+        RefillBuffer:
+            int i = totalCharsConsumed + (bytesLeftInBuffer * 3);
+
+        ReadByteFromInput:
+            if ((uint)(i + 2) >= length || input[i] != '%')
+                goto NoMoreOrInvalidInput;
+
+            uint value = input[i + 1];
+            if ((uint)((value - 'A') & ~0x20) <= ('F' - 'A'))
+            {
+                value = (value | 0x20) - 'a' + 10;
+            }
+            else if ((value - '8') <= ('9' - '8'))
+            {
+                value -= '0';
+            }
+            else goto NoMoreOrInvalidInput; // First character wasn't hex or was <= 7F (Ascii)
+
+            uint second = input[i + 2];
+            if ((second - '0') <= 9)
+            {
+                value = ((value << 4) + second) - '0';
+            }
+            else if ((uint)((second - 'A') & ~0x20) <= ('F' - 'A'))
+            {
+                value = ((value << 4) + (second | 0x20)) - 'a' + 10;
+            }
+            else goto NoMoreOrInvalidInput; // Second character wasn't Hex
+
+            Debug.Assert(value >= 128);
+
+            if (BitConverter.IsLittleEndian)
+            {
+                fourByteBuffer = (fourByteBuffer >> 8) | (value << 24);
+            }
+            else
+            {
+                fourByteBuffer = (fourByteBuffer << 8) | value;
+            }
+
+            if (++bytesLeftInBuffer != 4)
+            {
+                i += 3;
+                goto ReadByteFromInput;
+            }
+
+        DecodeRune:
+            Debug.Assert(totalCharsConsumed % 3 == 0);
+            Debug.Assert(bytesLeftInBuffer == 2 || bytesLeftInBuffer == 3 || bytesLeftInBuffer == 4);
+            Debug.Assert((fourByteBuffer & (BitConverter.IsLittleEndian ? 0x000000FF : 0xFF000000)) >= 128);
+            Debug.Assert((fourByteBuffer & (BitConverter.IsLittleEndian ? 0x0000FF00 : 0x00FF0000)) >= 128);
+            Debug.Assert(bytesLeftInBuffer < 3 || (fourByteBuffer & (BitConverter.IsLittleEndian ? 0x00FF0000 : 0x0000FF00)) >= 128);
+            Debug.Assert(bytesLeftInBuffer < 4 || (fourByteBuffer & (BitConverter.IsLittleEndian ? 0xFF000000 : 0x000000FF)) >= 128);
+
+            if (Rune.DecodeFromUtf8(fourByteSpan.Slice(0, bytesLeftInBuffer), out Rune rune, out bytesConsumed) == OperationStatus.Done)
+            {
+                Debug.Assert(bytesConsumed >= 2);
```

Comment:
It is theoretically possible to violate this assertion. I'm having trouble following the logic of this method so I don't know if an assertion violation will end up possibly buffer overrunning or otherwise having a negative effect.

```txt
Entry to method:

char* input = "%FA%FB%00"
fourByteBuffer = 0
bytesLeftInBuffer = 0
totalCharsConsumed = 0
charsToCopy = 0
bytesConsumed = 0

<after a few rounds of ReadByteFromInput>

fourByteBuffer = 0xFBFA0000 (buffer = [ 00 00 FA FB ])
bytesLeftInBuffer = 2

<go to NoMoreOrInvalidInput>

fourByteBuffer = 0x0000FBFA (buffer = [ FA FB 00 00 ])
bytesLeftInBuffer = 2

<go to DecodeRune>

DecodeFromUtf8 = InvalidData
bytesConsumed = 1
charsToCopy = 3

<go to AfterDecodeRune>

bytesLeftInBuffer = 1
totalCharsConsumed = 3

## meanwhile, another thread changes 'input' to read "%FA%FB%FC" ##

<go to RefillBuffer>

i = 3 + (1 * 3) = 6

<go to ReadByteFromInput>

fourByteBuffer = 0xFC0000FB (buffer = [ FB 00 00 FC ])
bytesLeftInBuffer = 2

<go to NoMoreOrInvalidInput>

## recall: bytesConsumed is still 1 from earlier DecodeRune call

fourByteBuffer = 0x00FC0000 (buffer = [ 00 00 FC 00 ])
bytesLeftInBuffer = 2

<go to DecodeRune>

DecodeFromUtf8 = Done
bytesConsumed = 1
Debug.Assert(bytesConsumed >= 2); // assertion would fail but not present in release branch

dest.Append(input + 3 - 3, 3); // copy 3 chars
charsToCopy = 0
charsToCopy = 3

<go to AfterDecodeRune>

...
```

Reply:
I believe there would be no negative side-effects in this case. While the assert would fail if present, at least one byte will be read by `Rune.DecodeFromUtf8` per iteration, so the *loop* will eventually exit. `bytesConsumed` being 1 instead of `>= 2` will not impact memory accesses into the input.
I can remove this assert or add a comment about this.

In general, is the scenario of a string instance being modified after creation something that we should be/are considering? I would be surprised if there aren't other APIs across runtime that make the assumption of string immutability with worse side-effects.